File: workflows/ingest/src/embed.ts
import { createHash } from 'crypto';
export const DIM = 768;

function tokenize(text: string): string[] {
  return text.toLowerCase().split(/\s+/).filter(Boolean);
}

function signedBucket(token: string): [number, number] {
  const h = createHash('sha256').update(token).digest();
  const idx = h.readUInt32BE(0) % DIM;
  const sign = (h[4] & 0x80) ? 1.0 : -1.0;
  return [idx, sign];
}

export function embed(text: string): number[] {
  const vec = new Array<number>(DIM).fill(0);
  for (const tok of tokenize(text)) {
    const [i, s] = signedBucket(tok);
    vec[i] += s;
  }
  const norm = Math.sqrt(vec.reduce((a,b)=>a+b*b, 0)) || 1.0;
  return vec.map(v => v / norm);
}

File: services/retrieval/app/embed.py
from hashlib import sha256
import math

DIM = 768


def _tokenize(text: str):
    return [t for t in text.lower().split() if t]


def _signed_bucket(token: str):
    h = sha256(token.encode("utf-8")).digest()
    idx = int.from_bytes(h[:4], "big") % DIM
    sign = 1.0 if (h[4] & 0x80) else -1.0
    return idx, sign


def embed(text: str):
    vec = [0.0] * DIM
    for tok in _tokenize(text):
        idx, sign = _signed_bucket(tok)
        vec[idx] += sign
    norm = math.sqrt(sum(v * v for v in vec)) or 1.0
    return [v / norm for v in vec]
